src/async.rs

```diff
@@ -269,3 +275,295 @@ impl ConnectionLike for Connection {
+                                self.send_state = SendState::Sent
+                            }
+                            Ok(Async::Ready(Some((item, senders)))) => {
+                                self.in_flight.push_back(senders);
+                                self.send_state = SendState::Unsent(item);
+                            }
+                        }
+                    }
+                    AsyncSink::NotReady(item) => {
+                        self.send_state = SendState::Unsent(item);
+                        return Ok(Async::NotReady);
+                    }
+                },
+                SendState::Sent => {
+                    match self.sink_stream.poll_complete()? {
+                        Async::Ready(()) => (),
+                        Async::NotReady => {
+                            self.send_state = SendState::Sent;
+                            return Ok(Async::NotReady);
+                        }
+                    }
+                    self.send_state = SendState::Free;
+                    return Ok(Async::Ready(Some(())));
+                }
+                SendState::Free => {
+                    let (item, senders) = match self.incoming.poll() {
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Ok(Async::Ready(Some(x))) => x,
+                        Ok(Async::Ready(None)) | Err(()) => return Ok(Async::Ready(None)),
+                    };
+                    self.in_flight.push_back(senders);
+                    self.send_state = SendState::Unsent(item);
+                }
+            }
+        }
+    }
+}
+
+impl<T> Pipeline<T>
+where
+    T: Sink<SinkError = <T as Stream>::Error> + Stream + Send + 'static,
+    T::SinkItem: Send,
+    T::Item: Send,
+    T::Error: Send,
+    T::Error: ::std::fmt::Debug,
+{
+    fn new(sink_stream: T) -> Self {
+        let (sender, receiver) = mpsc::unbounded();
+        tokio_executor::spawn(PipelineFuture {
+            sink_stream,
+            in_flight: VecDeque::new(),
+            senders: Vec::new(),
+            send_state: SendState::Free,
+            incoming: receiver,
+        });
+        Pipeline(sender)
+    }
+
+    // `None` means that the stream was out of items causing that poll loop to shut down.
+    fn send(
+        &self,
+        item: T::SinkItem,
+    ) -> Box<Future<Item = T::Item, Error = Option<T::Error>> + Send> {
+        Box::new(
+            self.send_recv_multiple(item, 1)
+                .into_future()
+                .map(|(item, _)| item.unwrap())
+                .map_err(|(err, _)| err),
+        )
+    }
+
+    fn send_recv_multiple(
+        &self,
+        item: T::SinkItem,
+        count: usize,
+    ) -> Box<Stream<Item = T::Item, Error = Option<T::Error>> + Send> {
+        let self_ = self.0.clone();
+
+        let mut senders = Vec::new();
+        let mut receivers = Vec::new();
```

Comment:
:+1: Thought about using https://github.com/servo/rust-smallvec here since 1 element vectors will be common. Seems like premature optimization though.